mlir/lib/Conversion/TosaToMLProgram/TosaToMLProgramPass.cpp

```diff
@@ -0,0 +1,49 @@
+//===- TosaToMLProgramPass.cpp - Lowering Tosa to MLProgram Dialect--------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass legalizes the TOSA dialect to the MLProgram dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/TosaToMLProgram/TosaToMLProgram.h"
+#include "mlir/Dialect/MLProgram/IR/MLProgram.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Transforms/Passes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_TOSATOMLPROGRAM
+#include "mlir/Conversion/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+using namespace tosa;
+
+namespace {
+struct TosaToMLProgram : public impl::TosaToMLProgramBase<TosaToMLProgram> {
+public:
+  void runOnOperation() override {
+    auto *context = &getContext();
+    auto moduleOp = getOperation();
+
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    target.addLegalDialect<ml_program::MLProgramDialect>();
+    target.addIllegalOp<tosa::VariableOp, tosa::VariableReadOp,
```

Comment:
I think only this setting may be needed and you could remove the other legal and dynamically legal settings as you are.using a partial conversion.

Reply:
Back from some experiments, seems we can not remove both `target.addLegalDialect<ml_program::MLProgramDialect>();` and `target.markUnknownOpDynamicallyLegal([](Operation *) { return true; });` 

Error: 
```
staging/llvm-project/mlir/test/Conversion/TosaToMLProgram/tosa-to-mlprogram.mlir:4 offset :5:3: error: failed to legalize operation 'tosa.variable' that was explicitly marked illegal
```

Am I missing something here?

Reply:
I'll check here, there should be something simpler. But not blocking.